File: utils/formatters.py
def fmt_price(v):
    if v is None:
        return "N/A"
    return f"${v:,.2f}"


def fmt_cap(v):
    if v is None:
        return "N/A"
    if v >= 1e12:
        return f"${v/1e12:.2f}T"
    if v >= 1e9:
        return f"${v/1e9:.2f}B"
    if v >= 1e6:
        return f"${v/1e6:.2f}M"
    return f"${v:,.0f}"


def fmt_pe(v):
    if v is None:
        return "N/A"
    return f"{v:.2f}x"


def register_filters(app):
    app.add_template_filter(fmt_price, "fmt_price")
    app.add_template_filter(fmt_cap, "fmt_cap")
    app.add_template_filter(fmt_pe, "fmt_pe")

File: frontend/src/utils/formatters.js
export function fmtPrice(v) {
  if (v == null) return 'N/A'
  return `$${v.toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`
}

export function fmtCap(v) {
  if (v == null) return 'N/A'
  if (v >= 1e12) return `$${(v / 1e12).toFixed(2)}T`
  if (v >= 1e9) return `$${(v / 1e9).toFixed(2)}B`
  if (v >= 1e6) return `$${(v / 1e6).toFixed(2)}M`
  return `$${v.toLocaleString()}`
}

export function fmtPE(v) {
  if (v == null) return 'N/A'
  return `${v.toFixed(2)}x`
}
